File: internal/maintenance/cat_plugins.go
package maintenance

import (
	"context"
	"path/filepath"
)

func init() { registerMatcher("plugins", 0, scanPlugins) } // no age gate: staleness = disabled

// scanPlugins surfaces reclaimable plugin storage under <root>/plugins:
//   - cache/<marketplace>/<plugin>/  — per-plugin cached repo (two levels deep),
//     cross-referenced against enabled plugins so a disabled/uninstalled plugin's
//     cache is flagged as a prime reclaim candidate and an enabled plugin's cache
//     is flagged safe-but-forces-redownload.
//   - marketplaces/<name>/  — per-marketplace metadata.
//   - repos/<name>/         — per cached repo.
//
// The matcher scopes strictly to cache/marketplaces/repos; plugins/ also holds
// data/, *.json and marker files that are config, not reclaimable cache. No age
// gate — plugin cache is regenerable, staleness is decided by enabled-state.
func scanPlugins(ctx context.Context, spec CategorySpec) ([]Candidate, error) {
	pluginsDir := filepath.Join(spec.Root, "plugins")
	enabled := make(map[string]bool, len(spec.Enabled))
	for _, k := range spec.Enabled {
		enabled[k] = true
	}

	out := []Candidate{}

	cache, err := scanPluginCache(ctx, filepath.Join(pluginsDir, "cache"), enabled)
	if err != nil {
		return nil, err
	}
	out = append(out, cache...)

	markets, err := scanPluginChildren(ctx, filepath.Join(pluginsDir, "marketplaces"), "marketplaces", "marketplace metadata")
	if err != nil {
		return nil, err
	}
	out = append(out, markets...)

	repos, err := scanPluginChildren(ctx, filepath.Join(pluginsDir, "repos"), "repos", "cached repo")
	if err != nil {
		return nil, err
	}
	out = append(out, repos...)

	// Anomaly: repos/ empty while cache/ holds data (audit found this broken/
	// legacy layout). Surface it as an informational flag on the cache
	// candidates — never auto-delete anything on account of it.
	if len(repos) == 0 && len(cache) > 0 {
		reposEntries, ok, _ := openDirNoSymlink(filepath.Join(pluginsDir, "repos"))
		if ok && len(reposEntries) == 0 {
			for i := range out {
				if out[i].Group == "cache" {
					stampMeta(&out[i], "layoutAnomaly", "repos-empty")
				}
			}
		}
	}

	return out, nil
}

// scanPluginCache walks cache/<marketplace>/<plugin>, one candidate per plugin.
func scanPluginCache(ctx context.Context, cacheDir string, enabled map[string]bool) ([]Candidate, error) {
	markets, ok, err := openDirNoSymlink(cacheDir)
	if err != nil || !ok {
		return []Candidate{}, err
	}

	out := []Candidate{}
	for _, market := range markets {
		if !market.IsDir() {
			continue
		}
		marketDir := filepath.Join(cacheDir, market.Name())
		plugins, ok, err := openDirNoSymlink(marketDir)
		if err != nil {
			return nil, err
		}
		if !ok {
			continue
		}
		for _, plugin := range plugins {
			if err := ctx.Err(); err != nil {
				return nil, err
			}
			if !plugin.IsDir() {
				continue
			}
			pluginDir := filepath.Join(marketDir, plugin.Name())
			bytes, files, newest, err := subtreeStats(ctx, pluginDir)
			if err != nil {
				return nil, err
			}
			isEnabled := enabled[plugin.Name()+"@"+market.Name()] || enabled[plugin.Name()]
			reason := "cached data for a disabled or uninstalled plugin"
			if isEnabled {
				reason = "cached data for an enabled plugin — safe to remove but forces a re-download on next use"
			}
			c := Candidate{
				Path: pluginDir, Bytes: bytes, Files: files, ModTime: newest,
				Reason: reason, Group: "cache",
				Meta: map[string]string{
					"marketplace": market.Name(),
					"plugin":      plugin.Name(),
					"enabled":     boolStr(isEnabled),
				},
			}
			out = append(out, c)
		}
	}
	return out, nil
}

// scanPluginChildren emits one candidate per immediate child dir of dir.
func scanPluginChildren(ctx context.Context, dir, group, reason string) ([]Candidate, error) {
	entries, ok, err := openDirNoSymlink(dir)
	if err != nil || !ok {
		return []Candidate{}, err
	}

	out := []Candidate{}
	for _, e := range entries {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		if !e.IsDir() {
			continue
		}
		child := filepath.Join(dir, e.Name())
		bytes, files, newest, err := subtreeStats(ctx, child)
		if err != nil {
			return nil, err
		}
		out = append(out, Candidate{
			Path: child, Bytes: bytes, Files: files, ModTime: newest,
			Reason: reason, Group: group,
			Meta: map[string]string{"name": e.Name()},
		})
	}
	return out, nil
}

func stampMeta(c *Candidate, k, v string) {
	if c.Meta == nil {
		c.Meta = map[string]string{}
	}
	c.Meta[k] = v
}

func boolStr(b bool) string {
	if b {
		return "true"
	}
	return "false"
}

File: src-tauri/src/maintenance/cat_plugins.rs
//! Ported from `internal/maintenance/cat_plugins.go` (W13). Surfaces reclaimable
//! plugin storage under `<root>/plugins`: per-plugin `cache/` (enabled
//! cross-referenced), `marketplaces/`, and `repos/`. No age gate — plugin cache
//! is regenerable; staleness is decided by enabled-state.

use std::collections::{BTreeMap, HashSet};
use std::path::Path;

use super::category::{bool_str, open_dir_no_symlink, subtree_stats};
use super::types::{Candidate, CategorySpec};

pub fn scan_plugins(spec: &CategorySpec) -> Result<Vec<Candidate>, String> {
    let plugins_dir = Path::new(&spec.root).join("plugins");
    let enabled: HashSet<&str> = spec.enabled.iter().map(String::as_str).collect();

    let mut out: Vec<Candidate> = Vec::new();

    let cache = scan_plugin_cache(&plugins_dir.join("cache"), &enabled)?;
    let cache_len = cache.len();
    out.extend(cache);

    let markets =
        scan_plugin_children(&plugins_dir.join("marketplaces"), "marketplaces", "marketplace metadata")?;
    out.extend(markets);

    let repos = scan_plugin_children(&plugins_dir.join("repos"), "repos", "cached repo")?;
    let repos_len = repos.len();
    out.extend(repos);

    // Anomaly: repos/ empty while cache/ holds data. Surface it as an
    // informational flag on the cache candidates — never auto-delete on it.
    if repos_len == 0 && cache_len > 0 {
        let (repos_entries, ok) = open_dir_no_symlink(&plugins_dir.join("repos"))?;
        if ok && repos_entries.is_empty() {
            for c in out.iter_mut() {
                if c.group == "cache" {
                    c.meta.insert("layoutAnomaly".to_string(), "repos-empty".to_string());
                }
            }
        }
    }

    Ok(out)
}

/// Walks `cache/<marketplace>/<plugin>`, one candidate per plugin. Mirrors Go
/// `scanPluginCache`.
fn scan_plugin_cache(cache_dir: &Path, enabled: &HashSet<&str>) -> Result<Vec<Candidate>, String> {
    let (markets, ok) = open_dir_no_symlink(cache_dir)?;
    if !ok {
        return Ok(Vec::new());
    }

    let mut out: Vec<Candidate> = Vec::new();
    for market in &markets {
        if !market.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            continue;
        }
        let market_name = market.file_name().to_string_lossy().into_owned();
        let market_dir = cache_dir.join(&market_name);
        let (plugins, ok) = open_dir_no_symlink(&market_dir)?;
        if !ok {
            continue;
        }
        for plugin in &plugins {
            if !plugin.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                continue;
            }
            let plugin_name = plugin.file_name().to_string_lossy().into_owned();
            let plugin_dir = market_dir.join(&plugin_name);
            let (bytes, files, newest) = subtree_stats(&plugin_dir);
            let is_enabled = enabled.contains(format!("{plugin_name}@{market_name}").as_str())
                || enabled.contains(plugin_name.as_str());
            let reason = if is_enabled {
                "cached data for an enabled plugin — safe to remove but forces a re-download on next use"
            } else {
                "cached data for a disabled or uninstalled plugin"
            };
            let mut meta = BTreeMap::new();
            meta.insert("marketplace".to_string(), market_name.clone());
            meta.insert("plugin".to_string(), plugin_name.clone());
            meta.insert("enabled".to_string(), bool_str(is_enabled).to_string());
            out.push(Candidate {
                path: plugin_dir.to_string_lossy().into_owned(),
                bytes,
                files,
                mod_time: newest,
                reason: reason.to_string(),
                group: "cache".to_string(),
                meta,
            });
        }
    }
    Ok(out)
}

/// Emits one candidate per immediate child dir of `dir`. Mirrors Go
/// `scanPluginChildren`.
fn scan_plugin_children(dir: &Path, group: &str, reason: &str) -> Result<Vec<Candidate>, String> {
    let (entries, ok) = open_dir_no_symlink(dir)?;
    if !ok {
        return Ok(Vec::new());
    }

    let mut out: Vec<Candidate> = Vec::new();
    for e in &entries {
        if !e.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            continue;
        }
        let name = e.file_name().to_string_lossy().into_owned();
        let child = dir.join(&name);
        let (bytes, files, newest) = subtree_stats(&child);
        let mut meta = BTreeMap::new();
        meta.insert("name".to_string(), name);
        out.push(Candidate {
            path: child.to_string_lossy().into_owned(),
            bytes,
            files,
            mod_time: newest,
            reason: reason.to_string(),
            group: group.to_string(),
            meta,
        });
    }
    Ok(out)
}

#[cfg(test)]
#[path = "cat_plugins_tests.rs"]
mod cat_plugins_tests;
